File: litfass-server/src/test/scala/lit/fass/server/script/GroovyScriptEngineTest.scala
package lit.fass.server.script

import java.util.Collections.singletonMap

import lit.fass.server.ScriptLanguage.GROOVY
import lit.fass.server.helper.UnitTest
import lit.fass.server.model.Data
import org.junit.experimental.categories.Category
import org.junit.runner.RunWith
import org.scalatest.matchers.should.Matchers
import org.scalatest.wordspec.AnyWordSpec
import org.scalatestplus.junit.JUnitRunner

import scala.jdk.CollectionConverters._

/**
 * @author Michael Mair
 */
@RunWith(classOf[JUnitRunner])
@Category(Array(classOf[UnitTest]))
class GroovyScriptEngineTest extends AnyWordSpec with Matchers {

  val scriptEngine: GroovyScriptEngine = GroovyScriptEngine()

  "groovy script engine" should {

    "be applicable to groovy" in {
      scriptEngine.isApplicable(GROOVY) shouldBe true
    }

    "return expected result" in {
      val script = """[bar: binding["data"]]"""
      val input = singletonMap("foo", Integer.valueOf(1))
      scriptEngine.invoke(script, Data.of(input).toList).get(0) shouldBe Map("bar" -> Map("foo" -> 1).asJava).asJava
    }

  }
}

File: litfass-server/src/test/kotlin/lit/fass/server/groovy/GroovyScriptEngineTest.kt
package lit.fass.server.groovy

import lit.fass.server.helper.TestTypes.UnitTest
import lit.fass.server.script.ScriptLanguage
import lit.fass.server.script.groovy.GroovyScriptEngine
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Tag
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.CsvSource

/**
 * @author Michael Mair
 */
@Tag(UnitTest)
internal class GroovyScriptEngineTest {

    val scriptEngine = GroovyScriptEngine()

    @ParameterizedTest(name = "{displayName} - {0}")
    @CsvSource(
        "KOTLIN, false",
        "GROOVY, true"
    )
    fun `groovy script engine is applicable`(language: String, expected: Boolean) {
        assertThat(scriptEngine.isApplicable(ScriptLanguage.valueOf(language))).isEqualTo(expected)
    }

    @Test
    fun `groovy script engine returns result`() {
        val script = """[bar: binding["data"]]"""
        val input = mapOf("foo" to 1)
        assertThat(scriptEngine.invoke(script, listOf(input)).first()).isEqualTo(mapOf("bar" to mapOf("foo" to 1)))
    }

    @Test
    fun `groovy script engine without a result`() {
        val script = """println "bar""""
        val input = mapOf("foo" to 1)
        assertThat(scriptEngine.invoke(script, listOf(input))).isEmpty()
    }
}